csrc/evaluator_common.cpp

```diff
@@ -334,15 +334,20 @@ void PrecomputedValues::bindTensorMetaData(
       "Something went wrong configuring launch. Inputs do not match.");
 
   for (const auto dim : c10::irange(logical_domain.size())) {
-    auto value = tensor.size((int64_t)dim);
-    if (logical_domain[dim]->hasExpandedExtent()) {
-      auto extent = logical_domain[dim]->extent();
-      auto expanded_extent = logical_domain[dim]->expandedExtent();
+    IterDomain* id = logical_domain[dim];
+    auto dim_size = tensor.size(static_cast<int64_t>(dim));
+    if (id->isDeviceDim()) {
+      dim_size = tv->getDeviceMesh().size();
+    }
```

Comment:
This is the only non-cosmetic change in this file. Bind the unsharded size when the dim is sharded.

Reply:
No, this would break. We are lacking APIs for multi-dimensional meshes, which https://github.com/NVIDIA/Fuser/pull/2260 tried to add and got deprioritized. However, I see your concern that this `getDeviceMesh().size()` call will **quietly** the program when we start to support multi-dimensional meshes. I don't have a great solution for that. 

cc @cowanmeg

Reply:
I think there's aspects of that PR we can pull out to make the DeviceMesh a bit more future proof. 
In the future we need something like getDeviceMesh.size(DIDx|DIDy|DIDz) that returns the size of the mesh in one dimension. 
We can pre-emptively parameterize DeviceMesh::size with parallel type, then we only have to update size's definition when multi-dimensional meshes are supported.

Reply:
Something like fail-safe assertions would be enough for now, but I think we should make sure to avoid silent failures.

Reply:
> Something like fail-safe assertions would be enough for now, but I think we should make sure to avoid silent failures.

agreed

> In the future we need something like getDeviceMesh.size(DIDx|DIDy|DIDz) that returns the size of the mesh in one dimension. We can pre-emptively parameterize DeviceMesh::size with parallel type, then we only have to update size's definition when multi-dimensional meshes are supported.

Other solution: `DeviceMesh::size` could return a tuple of ints. This has the advantage of not involving the parallel types and let us have an arbitrary number of dimensions for DeviceMeshes